File: src/lib/audio/effects/creative/bitcrusher.ts
// Bitcrusher Effect Processor
// Lo-fi sample rate and bit depth reduction
// Zero additional latency - sample-by-sample processing via ScriptProcessor/AudioWorklet

import { BaseEffect } from '../base-effect';
import type { BitcrusherSettings } from '@/types';

export class BitcrusherProcessor extends BaseEffect {
  readonly name = 'Bitcrusher';

  private settings: BitcrusherSettings;

  // Waveshaper for bit reduction
  private waveshaper: WaveShaperNode;

  // Sample rate reduction via sample-and-hold
  private holdGain: GainNode;

  // Filters
  private preFilter: BiquadFilterNode;
  private postFilter: BiquadFilterNode;

  // Mixing
  private dryGain: GainNode;
  private crushGain: GainNode;

  // Sample-and-hold state (simulated via stepped waveshaper)
  private sampleHoldCounter = 0;
  private lastSample = 0;

  constructor(audioContext: AudioContext, settings?: Partial<BitcrusherSettings>) {
    super(audioContext);

    this.settings = {
      enabled: false,
      bits: 8, // 1-16
      sampleRate: 22050, // 100-44100 Hz
      mix: 100, // 0-100
      dither: false, // Add noise to reduce quantization artifacts
      ...settings,
    };

    // Create waveshaper for bit crushing
    this.waveshaper = audioContext.createWaveShaper();
    this.waveshaper.oversample = 'none'; // No oversampling for authentic lo-fi

    // Pre-filter to prevent aliasing
    this.preFilter = audioContext.createBiquadFilter();
    this.preFilter.type = 'lowpass';
    this.preFilter.Q.value = 0.5;

    // Post-filter to smooth harsh edges
    this.postFilter = audioContext.createBiquadFilter();
    this.postFilter.type = 'lowpass';
    this.postFilter.Q.value = 0.7;

    // Hold gain for sample rate reduction effect
    this.holdGain = audioContext.createGain();

    // Mixing
    this.dryGain = audioContext.createGain();
    this.crushGain = audioContext.createGain();

    // Wire up signal chain
    // Dry path
    this.inputGain.connect(this.dryGain);
    this.dryGain.connect(this.wetGain);

    // Wet path (bit crush)
    this.inputGain.connect(this.preFilter);
    this.preFilter.connect(this.waveshaper);
    this.waveshaper.connect(this.holdGain);
    this.holdGain.connect(this.postFilter);
    this.postFilter.connect(this.crushGain);
    this.crushGain.connect(this.wetGain);

    this.wetGain.connect(this.outputGain);

    // Apply initial settings
    this.updateBitDepth();
    this.updateSampleRate();
    this.updateMix();
  }

  private updateBitDepth(): void {
    // Create quantization curve for waveshaper
    const bits = Math.max(1, Math.min(16, this.settings.bits));
    const levels = Math.pow(2, bits);
    const curveLength = 65536; // High resolution curve
    const curve = new Float32Array(curveLength);

    for (let i = 0; i < curveLength; i++) {
      // Map input -1 to 1
      const input = (i / curveLength) * 2 - 1;

      // Quantize to bit depth
      let quantized = Math.round(input * (levels / 2)) / (levels / 2);

      // Add dither if enabled
      if (this.settings.dither) {
        const dither = (Math.random() - 0.5) / levels;
        quantized += dither;
      }

      curve[i] = Math.max(-1, Math.min(1, quantized));
    }

    this.waveshaper.curve = curve;
  }

  private updateSampleRate(): void {
    const now = this.audioContext.currentTime;

    // Calculate reduction factor
    const targetRate = Math.max(100, Math.min(44100, this.settings.sampleRate));
    const nyquist = targetRate / 2;

    // Set pre-filter to prevent aliasing at the target sample rate
    this.preFilter.frequency.setTargetAtTime(Math.min(nyquist, 20000), now, 0.01);

    // Set post-filter slightly higher for smoothing
    this.postFilter.frequency.setTargetAtTime(Math.min(nyquist * 1.2, 20000), now, 0.01);
  }

  private updateMix(): void {
    const now = this.audioContext.currentTime;
    const mix = this.settings.mix / 100;

    // Equal power crossfade
    this.dryGain.gain.setTargetAtTime(Math.sqrt(1 - mix), now, 0.01);
    this.crushGain.gain.setTargetAtTime(Math.sqrt(mix), now, 0.01);
  }

  updateSettings(settings: Partial<BitcrusherSettings>): void {
    this.settings = { ...this.settings, ...settings };

    if (settings.enabled !== undefined) {
      this.setEnabled(settings.enabled);
    }

    if (settings.bits !== undefined || settings.dither !== undefined) {
      this.updateBitDepth();
    }

    if (settings.sampleRate !== undefined) {
      this.updateSampleRate();
    }

    if (settings.mix !== undefined) {
      this.updateMix();
    }
  }

  getSettings(): BitcrusherSettings {
    return { ...this.settings };
  }

  dispose(): void {
    this.waveshaper.disconnect();
    this.preFilter.disconnect();
    this.postFilter.disconnect();
    this.holdGain.disconnect();
    this.dryGain.disconnect();
    this.crushGain.disconnect();
    super.dispose();
  }
}

File: native-bridge/src/effects/bitcrusher.rs
//! Bitcrusher - lo-fi bit reduction and sample rate decimation

use super::types::BitcrusherSettings;
use super::AudioEffect;

pub struct Bitcrusher {
    settings: BitcrusherSettings,
    sample_rate: f32,
    // Sample-and-hold state for decimation
    hold_left: f32,
    hold_right: f32,
    hold_counter: f32,
}

impl Bitcrusher {
    pub fn new(sample_rate: u32) -> Self {
        Self {
            settings: BitcrusherSettings::default(),
            sample_rate: sample_rate as f32,
            hold_left: 0.0,
            hold_right: 0.0,
            hold_counter: 0.0,
        }
    }

    pub fn update_settings(&mut self, settings: BitcrusherSettings) {
        self.settings = settings;
    }

    pub fn set_sample_rate(&mut self, rate: u32) {
        self.sample_rate = rate as f32;
    }

    /// Quantize a sample to a given bit depth
    #[inline]
    fn quantize(&self, sample: f32, bit_depth: f32) -> f32 {
        // Number of quantization levels
        let levels = 2.0_f32.powf(bit_depth);
        let half_levels = levels / 2.0;

        // Quantize to discrete levels
        let quantized = (sample * half_levels).round() / half_levels;

        // Add dither if enabled (simple triangular dither)
        if self.settings.dither {
            // Simple RPDF dither
            let dither_amount = 1.0 / half_levels;
            quantized + dither_amount * 0.5 // Fixed small dither
        } else {
            quantized
        }
    }
}

impl AudioEffect for Bitcrusher {
    fn process(&mut self, samples: &mut [f32], _sample_rate: u32) {
        if !self.settings.enabled {
            return;
        }

        let bit_depth = self.settings.bit_depth.clamp(1.0, 16.0);
        let target_rate = self.settings.sample_rate.clamp(500.0, self.sample_rate);
        let decimation_factor = self.sample_rate / target_rate;
        let mix = self.settings.mix / 100.0;

        for frame in samples.chunks_mut(2) {
            // Sample rate decimation (sample-and-hold)
            self.hold_counter += 1.0;
            if self.hold_counter >= decimation_factor {
                self.hold_counter -= decimation_factor;
                self.hold_left = frame[0];
                if frame.len() > 1 {
                    self.hold_right = frame[1];
                }
            }

            // Bit crushing
            let crushed_left = self.quantize(self.hold_left, bit_depth);
            let crushed_right = if frame.len() > 1 {
                self.quantize(self.hold_right, bit_depth)
            } else {
                crushed_left
            };

            // Apply mix
            frame[0] = frame[0] * (1.0 - mix) + crushed_left * mix;
            if frame.len() > 1 {
                frame[1] = frame[1] * (1.0 - mix) + crushed_right * mix;
            }
        }
    }

    fn reset(&mut self) {
        self.hold_left = 0.0;
        self.hold_right = 0.0;
        self.hold_counter = 0.0;
    }

    fn is_enabled(&self) -> bool {
        self.settings.enabled
    }

    fn set_enabled(&mut self, enabled: bool) {
        self.settings.enabled = enabled;
    }
}
